src/service-worker.ts

```diff
@@ -0,0 +1,125 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import {writeFile} from 'fs';
+import * as path from 'path';
+import * as logging from 'plylog';
+import {PolymerProject} from './polymer-project';
+import {DepsIndex} from './analyzer';
+import {SWConfig, generate as swPrecacheGenerate} from 'sw-precache';
+
+let logger = logging.getLogger('polymer-build.service-worker');
+
+export interface AddServiceWorkerOptions {
+  project: PolymerProject;
+  buildRoot: string;
+  bundled?: boolean;
+  serviceWorkerPath?: string;
+  swConfig?: SWConfig;
+}
+
+/**
+ * Parses a depsIndex object and returns an array of file paths, pointing to
+ * the project's required service worker precached assets.
+ */
+function getPrecachedAssets(depsIndex: DepsIndex, project: PolymerProject): string[] {
+  let precachedAssets = new Set<string>(project.analyzer.allFragments);
+  precachedAssets.add(project.entrypoint);
+
+  for (let dep of depsIndex.depsToFragments.keys()) {
+    precachedAssets.add(dep);
+  }
+  for (let depImports of depsIndex.fragmentToFullDeps.values()) {
+    depImports.scripts.forEach((s) => precachedAssets.add(s));
+    depImports.styles.forEach((s) => precachedAssets.add(s));
+  }
+  return Array.from(precachedAssets);
+}
+
+/**
+ * Parses a depsIndex object and returns an array of file paths for a BUNDLED
+ * build, pointing to the project's required service worker precached assets.
+ */
+function getBundledPrecachedAssets(project: PolymerProject) {
+  let precachedAssets = new Set<string>(project.analyzer.allFragments);
+  precachedAssets.add(project.entrypoint);
+
+  return Array.from(precachedAssets).concat(project.bundler.sharedBundleUrl);
```

Comment:
might as well add the sharedBundleUrl to the precachedAssest like you do the entrypoint, so it's consistent